### PR TITLE
Feature/Smart Detector Alert Rules

### LINKED ISSUE #4369: Feature: Support for creating Smart Detector Alert Rules through ASO
This is a request to add [Smart Detector Alert Rules](https://learn.microsoft.com/en-us/rest/api/monitor/smart-detector-alert-rules) in `Support Resources" list of ASO.

### PR DESCRIPTION
**Pull request by bingikarthik**
Closes #4369

**What this PR does / why we need it**:
Adding new resource microsoft.alertsmanagement/smartDetectorAlertRules

**Special notes for your reviewer**:

**How does this PR make you feel**: Quite exited!

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains tests
- [x] this PR contains YAML Samples

### CODE REVIEW

File: v2/specs/azure-rest-api-specs
Comment (theunrepentantgeek) [inline v2/specs/azure-rest-api-specs:1]: We need the submodule in order for the code generator to run; please revert this deletion.

Comment (bingikarthik) [reply v2/specs/azure-rest-api-specs:1]: Thanks, I have performed the rebase to branch as discussed over slack thread!

Comment (theunrepentantgeek) [reply v2/specs/azure-rest-api-specs:1]: Reactivating: GitHub still shows that you've deleted this submodule, breaking your branch. We can't troubleshoot your problems in this state.

Comment (bingikarthik) [reply v2/specs/azure-rest-api-specs:1]: Apologies, I thought i have added during rebase operation!
It's been added now. Please check

**Review by theunrepentantgeek — APPROVED** [review]
This looks solid, thanks!

As a side effect of the update to the `azure-rest-api-specs` submodule, I expect the CI build for this will fail in the test `Test_Networking_Route_CreatedThenRouteTableUpdated_RouteStillExists`. This will be resolved in PR #4386.

**Comment by theunrepentantgeek** [comment]
> I expect the CI build for this will fail in the test Test_Networking_Route_CreatedThenRouteTableUpdated_RouteStillExists.

To clarify, the upstream changes to remove the property `HasBgpOverride` cause the test recording to fail; I've remedied this in the linked PR.

File: v2/api/containerservice/v1api20240402preview/arm/managed_clusters_agent_pool_spec_types_gen.go [hunk 1]
```diff
@@ -648,8 +648,7 @@ type ScaleProfile struct {
 	// at most one AutoScaleProfile is allowed.
 	Autoscale []AutoScaleProfile `json:"autoscale"`
 
-	// Manual: Specifications on how to scale the VirtualMachines agent pool to a fixed size. Currently, at most one
-	// ManualScaleProfile is allowed.
+	// Manual: Specifications on how to scale the VirtualMachines agent pool to a fixed size.
```
[the remark this answers is not in the document]
Comment (bingikarthik) [reply v2/api/containerservice/v1api20240402preview/arm/managed_clusters_agent_pool_spec_types_gen.go:651]: Sure, will wait for the PR: https://github.com/Azure/azure-service-operator/pull/4386 gets merged and will do the rebase post that!

**Comment by bingikarthik** [comment]
> This looks solid, thanks!
> 
> As a side effect of the update to the `azure-rest-api-specs` submodule, I expect the CI build for this will fail in the test `Test_Networking_Route_CreatedThenRouteTableUpdated_RouteStillExists`. This will be resolved in PR #4386.

Thanks for the review, will wait for the PR #4386 to gets merged!

File: v2/azure-arm.yaml [hunk 2]
```diff
@@ -921,6 +921,19 @@ status:
 #
 objectModelConfiguration:
   alertsmanagement:
+    2021-04-01:
+      SmartDetectorAlertRule:
+        $export: true
+        $supportedFrom: v2.10.0
```
[the remark this answers is not in the document]
Comment (bingikarthik) [reply v2/azure-arm.yaml]: This was updated now!

**Comment by bingikarthik** [comment]
The PR failing on test-generator CI test due to following error:
```
while decoding ../../samples/alertsmanagement/v1api20210401/v1api20210401_smartdetectoralertrule.yaml: json: unknown field "scopesReferences"
```
This was handled now in following commit: https://github.com/Azure/azure-service-operator/pull/4375/commits/bd4487bee264cc2a159bcbd5fc9ef48264be61c4
Could you please re-trigger the tests again? TIA

**Comment by theunrepentantgeek** [comment]
Your fix for `v1api20210401_smartdetectoralertrule.yaml` worked; we're down to just one failing test.

The test `Test_Samples_CreationAndDeletion/Test_AlertsManagement_SmartDetectorAlertRules_CRUD` is verifying that your sample in `v2/samples/alertsmanagement/v1api20210401` works, and the test failure has me a little confused.

As recorded, it's failing with this error:

```
cannot find go-vcr recording for request from test "Test_Samples_CreationAndDeletion/Test_Alertsmanagement_v1api20210401_CreationAndDeletion" 
(cassette: "recordings/Test_Samples_CreationAndDeletion/Test_Alertsmanagement_v1api20210401_CreationAndDeletion")
(no responses recorded for this method/URL): 
PUT https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/asotest-rg-qiusii?api-version=2020-06-01  
(hash: 9cba6d48313722b9080a0dd069b5778157a9d8a220d387fddc6acccf803f140d)
```

This error is typical when the recording is out of date, and our normal approach is to just re-record the test, hand check the changes to the recording file, commit the changes and move on.

However, when I delete the recording and run the test in recording mode, I get an error I didn't anticipate: 

I'm running the test with this command:
``` bash
$ TEST_FILTER=Test_Samples_CreationAndDeletion/Test_Alertsmanagement_v1api20210401_CreationAndDeletion TIMEOUT=60m task controller:test-integration-envtest
```

And it's failing with this error (lightly edited for clarity):

``` text
kube_per_test_context.go:376: 
    Expected success, but got an error:
        <*errors.StatusError | 0xc004a47e00>: 
        SmartDetectorAlertRule.alertsmanagement.azure.com "asotestfjozrf" is invalid: [spec.actionGroups: Required value, spec.detector: Required value]
        {
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {
                    SelfLink: "",
                    ResourceVersion: "",
                    Continue: "",
                    RemainingItemCount: nil,
                },
                Status: "Failure",
                Message: "SmartDetectorAlertRule.alertsmanagement.azure.com \"asotestfjozrf\" is invalid: [spec.actionGroups: Required value, spec.detector: Required value]",
                Reason: "Invalid",
                Details: {
                    Name: "asotestfjozrf",
                    Group: "alertsmanagement.azure.com",
                    Kind: "SmartDetectorAlertRule",
                    UID: "",
                    Causes: [
                        {
                            Type: "FieldValueRequired",
                            Message: "Required value",
                            Field: "spec.actionGroups",
                        },
                        {
                            Type: "FieldValueRequired",
                            Message: "Required value",
                            Field: "spec.detector",
                        },
                    ],
                    RetryAfterSeconds: 0,
                },
                Code: 422,
            },
        }
```

This is the `kube-apiserver` complaining about missing mandatory fields. Reviewing the CRD code, this is correct:

``` go
	// +kubebuilder:validation:Required
	// ActionGroups: The alert rule actions.
	ActionGroups *ActionGroupsInformation `json:"actionGroups,omitempty"`
```
and
``` go
	// +kubebuilder:validation:Required
	// Detector: The alert rule's detector.
	Detector *Detector `json:"detector,omitempty"`
```

The fix will be to update the sample with the required fields and re-record the test. 

What's confusing to me is the question of how the recording was ever created in the first place? Do you have any insights?

**Comment by theunrepentantgeek** [comment]
That's a wall of text, apologies for that.

**TL;DR:** CI is failing in the test `Test_AlertsManagement_SmartDetectorAlertRules_CRUD` verifying your sample. Fix is to update the sample by adding the missing mandatory fields, rerecord the test, then push your changes.

**Comment by bingikarthik** [comment]
No problem @theunrepentantgeek, that's detailed explanation!
I will work on adding mandatory fields to the tests.

**Review by theunrepentantgeek — APPROVED** [review]

**Comment by theunrepentantgeek** [comment]
Failing due to a compilation error:

```
internal/controllers/alertsmanagement_smartdetectoralertrules_crud_v20210401_test.go:66:18: cannot use detectorId (variable of type *"github.com/Azure/azure-service-operator/v2/api/alertsmanagement/v1api20210401".Detector_STATUS) as *"github.com/Azure/azure-service-operator/v2/api/alertsmanagement/v1api20210401".Detector value in struct literal
```

Recommend you run `task ci` locally before pushing your next set of changes; this is the same command line as used by our CI tooling.

**Comment by bingikarthik** [comment]
> Failing due to a compilation error:
> 
> ```
> internal/controllers/alertsmanagement_smartdetectoralertrules_crud_v20210401_test.go:66:18: cannot use detectorId (variable of type *"github.com/Azure/azure-service-operator/v2/api/alertsmanagement/v1api20210401".Detector_STATUS) as *"github.com/Azure/azure-service-operator/v2/api/alertsmanagement/v1api20210401".Detector value in struct literal
> ```
> 
> Recommend you run `task ci` locally before pushing your next set of changes; this is the same command line as used by our CI tooling.

Thanks for sharing the error, this was caused due to following commit: https://github.com/Azure/azure-service-operator/pull/4375/commits/0df7cbe7ef25f202ecab476a8f8daf984f2e639a and it was reverted now!

File: v2/internal/controllers/alertsmanagement_smartdetectoralertrules_crud_v20210401_test.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	alertsmanagement "github.com/Azure/azure-service-operator/v2/api/alertsmanagement/v1api20210401"
+	insights "github.com/Azure/azure-service-operator/v2/api/insights/v1api20200202"
+	insightsag "github.com/Azure/azure-service-operator/v2/api/insights/v1api20230101"
+	"github.com/Azure/azure-service-operator/v2/internal/testcommon"
+	"github.com/Azure/azure-service-operator/v2/internal/util/to"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+func Test_AlertsManagement_SmartDetectorAlertRules_CRUD(t *testing.T) {
```
Comment (matthchr) [inline v2/internal/controllers/alertsmanagement_smartdetectoralertrules_crud_v20210401_test.go:22]: I think you're missing a recording for this test - there should be a `recordings/Test_AlertsManagement_SmartDetectorAlertRules_CRUD.yaml` file as well.

Can you run ` go test -count=1 -timeout 60m -run 'Test_AlertsManagement_SmartDetectorAlertRules_CRUD' -v ./internal/controllers`?

That should produce a new recording for this test and ensure it passes.

Comment (bingikarthik) [reply v2/internal/controllers/alertsmanagement_smartdetectoralertrules_crud_v20210401_test.go:22]: Thanks, changes has been pushed now!

**Review by matthchr — APPROVED** [review]

**Review by theunrepentantgeek — APPROVED** [review]
